File: 03-sliding_window/00567-Permutation_in_String/00567-permutation_in_string.py
# 567: Permutation in String
# https://leetcode.com/problems/permutation-in-string/


class Solution:
    # SOLUTION
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2): return False
        
        s1Count = [0]*26
        for c in s1: s1Count[ord(c)-ord('a')]+=1
        s2Count = [0]*26
        i, j = 0, 0
        
        while j < len(s2):
            s2Count[ord(s2[j])-ord('a')]+=1
            
            if j-i+1 == len(s1):
                if s1Count == s2Count:
                    return True
            if j-i+1 < len(s1):
                j+=1
            else:
                s2Count[ord(s2[i])-ord('a')]-=1
                i+=1
                j+=1


        return False



if __name__ == "__main__":
    o = Solution()

    # INPUT
    s1: str = "ab"
    s2: str = "eidbaooo"

    # OUTPUT
    result = o.checkInclusion(s1, s2)
    print(result)

File: 03-sliding_window/00567-Permutation_in_String/00567-permutation_in_string.java
// 567: Permutation in String
// https://leetcode.com/problems/permutation-in-string/


import java.util.Arrays;


class Solution {
    // SOLUTION
    boolean checkInclusion(String s1, String s2) {
        if (s1.length() > s2.length()) return false;
        
        int[] s1Count = new int[26]; for (var c : s1.toCharArray()) s1Count[c-'a']++;
        int[] s2Count = new int[26];
        int i = 0;
        int j = 0;
        
        while (j < s2.length()) {
            s2Count[s2.charAt(j)-'a']++;
            
            if (j-i+1 == s1.length()) 
                if (Arrays.equals(s1Count, s2Count)) 
                    return true;
            if (j-i+1 < s1.length()) 
                j++;
            else {
                s2Count[s2.charAt(i)-'a']--;
                i++;
                j++;
            }
        }

        return false;
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        String s1 = "ab";
        String s2 = "eidbaooo";

        // OUTPUT
        var result = o.checkInclusion(s1, s2);
        System.out.println(result);
    }
}